README.md

# sGBJ

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/sGBJ)](https://CRAN.R-project.org/package=sGBJ)
[![R-CMD-check](https://github.com/lauravillain/sGBJ/workflows/R-CMD-check/badge.svg)](https://github.com/lauravillain/sGBJ/actions)
<!-- badges: end -->

The goal of `sGBJ` is to provide an extension of the Generalized Berk-Jones test for set-based inference in genomics studies with a censored time-to-event outcome. In particular, it is geared towards 

This packages implements an extension of the GBJ statistic for censored time-to-event data. It computes the sGBJ statistic and its p-value for testing the association between a gene set and a time-to-event outcome with possible adjustment on additional covariates.

The main function of the package is `sGBJ()`

The methods implemented in this package are detailed in the following article:

> Villain L, Ferté T, Thiébaut R, Hejblum BP (2021). Gene Set Analysis for time-to-event outcome with the Generalized Berk–Jones statistic. *bioRxiv* 2021.09.07.459329.  
DOI: [10.1101/2021.09.07.459329](https://www.biorxiv.org/content/10.1101/2021.09.07.459329)

## Installation

You can install the released version of sGBJ from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("sGBJ")
```

Laura Villain, Thomas Ferté, Rodolphe Thiébaut & Boris Hejblum



NEWS.md
# News about the `sGBJ` R package

### Main changes in Version 0.1.0 (2021-09-06):

 * This is the first release


tests/testthat.R
library(testthat)
library(sGBJ)

test_check("sGBJ")


tests/testthat/test-sGBJ_scores.R
test_that("sGBJ_scores_consistency", {
  set.seed(1)
  n <- 10
  surv_data <- data.frame(Time = runif(n = n, min = 0, max = 100),
                          event = rbinom(n = n, size = 1, prob = 0.5))
  surv <- survival::Surv(time = surv_data$Time, event = surv_data$event)

  counts_pathway <- data.frame(P1 = rnorm(n = n),
                               P2 = rnorm(n = n))

  ls_results <- sGBJ::sGBJ_scores(surv,counts_pathway, nperm = 2)

  expect_equal(ls_results$test_stats, sGBJ::ls_test_results$test_stats)
  expect_equal(ls_results$cor_mat, sGBJ::ls_test_results$cor_mat)

})


R/sGBJ.R
#' Compute the sGBJ statistic and its p-value quantifying a gene set expression association
#' with survival
#'
#'This function is the main function of the sGBJ package to perform Gene Set Analysis
#'in the context of time-to-event outcome.
#'
#' @param surv a \code{\link[survival]{Surv}} object of length \code{n}
#' @param factor_matrix a \code{n x p} \code{data.frame} of the expression for the
#' particular gene set of interest being tested
#' @param covariates a \code{n x l} matrix of the covariates to adjust upon. Default is \code{NULL}
#' @param nperm number of permutations performed to estimate the \code{epsilon} matrix.
#' Default is \code{300}.
#'
#' @return The sGBJ statistic and its associated p-value associated
#' @examples
#' n <- 100
#' surv_data <- data.frame(Time = runif(n = n, min = 0, max = 100),
#'                         event = rbinom(n = n, size = 1, prob = 0.5))
#' surv <- survival::Surv(time = surv_data$Time, event = surv_data$event)
#'
#' factor_matrix <- data.frame(P1 = rnorm(n = n),
#'                              P2 = rnorm(n = n))
#'
#' sGBJ::sGBJ(surv,factor_matrix, nperm = 2)
#'
#' @export
#'
sGBJ <- function(surv, factor_matrix, covariates = NULL, nperm = 300){

  scores_GBJ <- sGBJ_scores(surv, factor_matrix, covariates, nperm)

  GBJOut <- GBJ::GBJ(test_stats = scores_GBJ$test_stats,
                     cor_mat = scores_GBJ$cor_mat)

  names(GBJOut) <- c("sGBJ_stat", "sGBJ_pvalue", "err_code")

  return(GBJOut)
}


R/epsilon_matrix.R
#' .epsilon_matrix
#'
#' @description Compute the epsilon matrix by permutation for the \code{sGBJ_scores()} function.
#'
#' @param Z the score vector returned by \code{.survival_scores()} function.
#' @param surv a \code{\link[survival]{Surv}} object of length \code{n}
#' @param factor_matrix a \code{n x p} \code{data.frame} of the expression for the
#' particular gene set of interest being tested
#' @param covariates a \code{n x l} matrix of the covariates to adjust upon. Default is \code{NULL}
#' @param nperm number of permutations performed to estimate the \code{epsilon} matrix.
#' Default is \code{300}.
#' @param dat data used to fit survival model returned by \code{.survival_scores()} function.
#'
#' @return The epsilon matrix.
.epsilon_matrix <- function(Z, nperm, surv, factor_matrix, covariates = NULL, dat){

  if(!is.null(covariates)){
    if (is.null(dim(covariates))){
      size_covariates <- 1
    }else{
      size_covariates <- ncol(covariates)
    }
  }

  # build Z_matrix
  Z_matrix <- matrix(nrow = (length(Z)), ncol = nperm)
  Z_matrix[,1] <-  Z
  i <- 2

  # fill Z_matrix
  while(i <= nperm){
    perm_OK <- TRUE
    perm <- sample(length(surv))
    surv_perm <- surv[perm]
    if(is.null(covariates)){
      dat_perm <- factor_matrix
    } else {
      if (size_covariates>1){
        covariates_perm <- covariates[perm,]
      }else{
        covariates_perm <- unlist(covariates)[perm]
      }
      dat_perm=cbind(covariates_perm, factor_matrix)
    }
    for (j in 1:(length(Z))){
      if(is.null(covariates)){
        model=try(survival::coxph(surv_perm ~ dat_perm[, j]))
      } else {
        vecCovariates <- paste("dat_perm[,", size_covariates, "]", collapse=" + ")
        vecPathway <- paste("dat_perm[,", j + size_covariates,"]", sep="")
        formX <- paste(c(vecPathway, vecCovariates), collapse = " + ")
        form <- as.formula(paste0("surv_perm ~ ", formX))
        model <- try(survival::coxph(form, data = dat_perm))
      }

      boolLengthModel <- length(model)>10
      if(boolLengthModel){
        if(is.null(covariates)){
          Z_matrix[j, i] <- model$coefficients / summary(model)$coefficients[3]
        } else {
          Z_matrix[j, i] <- model$coefficients[1] / summary(model)$coefficients[1,3]
        }
      }

      boolZna <- !is.na(Z_matrix[j,i])
      boolZinf <- abs(Z_matrix[j,i]) != Inf

      perm_OK <- boolLengthModel & boolZna & boolZinf
    }
    i <- i + perm_OK
  }

  epsilon <- cor(t(Z_matrix))

  return(epsilon)
}


R/survival_scores.R
#' .survival_scores
#'
#' @description Compute the survival score
#'
#' @param surv a \code{\link[survival]{Surv}} object of length \code{n}
#' @param factor_matrix a \code{n x p} \code{data.frame} of the expression for
#' the particular gene set of interest being tested
#' @param covariates a matrix nxl of the covariates to adjust. Default is \code{NULL}
#'
#' @importFrom stats as.formula
#'
#' @return A list of length 3 with the updated factor_matrix (same as factor_matrix but removing columns for which survival model failed to converge), the Z matrix and the data used to fit survival model.
.survival_scores <- function(factor_matrix, covariates = NULL, surv){
  remove_Z <- NULL
  Z <- numeric(ncol(factor_matrix))

  if(is.null(covariates)){
    dat <- factor_matrix
  } else {
    dat <- cbind(covariates, factor_matrix)
    if (is.null(dim(covariates))){
      size_covariates <- 1
    }else{
      size_covariates <- ncol(covariates)
    }
  }

  remove_Z <- rep(FALSE, ncol(factor_matrix))
  for (i in 1:ncol(factor_matrix)){

    if(is.null(covariates)){
      model <- try(survival::coxph(surv ~ dat[,i]))
    } else {
      vecCovariates <- paste("dat[,", size_covariates, "]", collapse=" + ")

      vecPathway <- paste("dat[,", i+size_covariates,"]", sep = "")
      formX <- paste(c(vecPathway, vecCovariates), collapse = " + ")
      form <- as.formula(paste0("surv ~ ", formX))

      model <- try(survival::coxph(form, data = dat))
    }

    boolLengthModel <- length(model) > 10
    if(boolLengthModel){
      if(is.null(covariates)){
        Z[i] <- model$coefficients / summary(model)$coefficients[3]
      } else {
        Z[i] <- model$coefficients[1] / summary(model)$coefficients[1, 3]
      }
    }

    boolZna <- is.na(Z[i])
    boolZinf <- abs(Z[i]) == Inf
    boolRemove <- (!boolLengthModel) | boolZna | boolZinf

    if(boolRemove){
      remove_Z[i] <- TRUE
    }
  }

  Z <- Z[!remove_Z]
  updatedFactor_matrix <-  factor_matrix[, !remove_Z]

  return(list(Z = Z,
              updatedFactor_matrix = updatedFactor_matrix,
              dat = dat))
}


R/surv_calc_scores_stats.R
#' surv_calc_scores_stats
#'
#' @description An adaptation of \code{GBJ::calc_scores_stats()} to survival context.
#' Wrapper of sGBJ_scores() function.
#'
#' @param null_model An R cox model fitted with \code{survival::coxph()}.
#' @param factor_matrix An \code{n x p} matrix with each factor as one column. There should be no missing data.
#' @param nperm Number of permutations (default is 300)
#'
#' @return A list with the elements:
#' \item{test_stats}{The \code{p} score test statistics.}
#' \item{cor_mat}{The \code{p x p} matrix giving the pairwise correlation of every test statistic pairs.}
#'
#' @export
#'
#' @examples
#' n <- 100
#' surv_data <- data.frame(Time = runif(n = n, min = 0, max = 100),
#'                         event = rbinom(n = n, size = 1, prob = 0.5))
#' surv <- survival::Surv(time = surv_data$Time, event = surv_data$event)
#'
#' factor_matrix <- data.frame(P1 = rnorm(n = n),
#'                              P2 = rnorm(n = n))
#'
#' covariates <- data.frame(age = runif(n = n, 60, 90))
#'
#' null_model <- survival::coxph(surv ~ age, data = covariates, x = TRUE)
#' surv_reg_stats <- surv_calc_scores_stats(null_model = null_model,
#'                                          factor_matrix = factor_matrix,
#'                                          nperm = 2)#nperm = 300)
#'
#' GBJ::GBJ(test_stats=surv_reg_stats$test_stats, cor_mat=surv_reg_stats$cor_mat)
#'
surv_calc_scores_stats <- function(null_model,
                                   factor_matrix,
                                   nperm = 300){

  # check null_model
  if(class(null_model) != "coxph") {
    stop("null_model must be a 'coxph' object from survival::coxph()")
  }
  if(!any(names(null_model) == "x")) {
    stop("Please be sure to provide the 'x = TRUE' option to survival::coxph() function")
  }

  res <- sGBJ_scores(surv = null_model$y,
                     factor_matrix = factor_matrix,
                     nperm = nperm,
                     covariates = null_model$x)

  return(res)
}


R/sGBJ_scores.R
#' Compute the sGBJ statistic along with its p-value quantifying the association
#' between a gene set and survival outcome
#'
#' @param surv a \code{\link[survival]{Surv}} object of length \code{n}
#' @param factor_matrix a \code{n x p} \code{data.frame} of the expression for the
#' particular gene set of interest being tested
#' @param covariates a \code{n x l} matrix of the covariates to adjust upon. Default is \code{NULL}
#' @param nperm number of permutations performed to estimate the \code{epsilon} matrix.
#' Default is \code{300}.
#'
#' @importFrom stats cor
#' @export
#'
#' @return a list containing the sGBJ statistic estimation and its associated p-value
#' @examples
#'  n <- 100
#'  surv_data <- data.frame(Time = runif(n = n, min = 0, max = 100),
#'                          event = rbinom(n = n, size = 1, prob = 0.5))
#'  surv <- survival::Surv(time = surv_data$Time, event = surv_data$event)
#'
#'  factor_matrix <- data.frame(P1 = rnorm(n = n),
#'                               P2 = rnorm(n = n))
#'
#'  sGBJ::sGBJ_scores(surv,factor_matrix, nperm = 2)
#'
#'  # with covariates
#'
#'  covariates <- data.frame(age = runif(n = n, 60, 90))
#'
#'  sGBJ_scores(surv,factor_matrix, nperm = 2, covariates = covariates)
sGBJ_scores=function(surv, factor_matrix, covariates = NULL, nperm = 300){

  # computation of the score vector
  lsScores <- .survival_scores(factor_matrix = factor_matrix,
                               covariates = covariates,
                               surv = surv)

  # computation of the epsilon matrix by permutation
  epsilon <- .epsilon_matrix(Z = lsScores$Z,
                             nperm = nperm,
                             surv = surv,
                             factor_matrix = lsScores$updatedFactor_matrix,
                             covariates = covariates,
                             dat = lsScores$datas)

  scores_GBJ <- list("test_stats" = lsScores$Z,
                     "cor_mat" = epsilon)

  return(scores_GBJ)
}

